src/Serializing/SExpression/Precise.hs

```diff
@@ -67,5 +67,16 @@ instance GToSExpression U1 where
 instance GToSExpression f => GToSExpression (M1 S s f) where
   gtoSExpression = gtoSExpression . unM1 -- FIXME: show the selector name, if any
 
-instance ToSExpression k => GToSExpression (K1 R k) where
-  gtoSExpression k = pure . toSExpression (unK1 k)
+instance Show k => GToSExpression (K1 R k) where
+  gtoSExpression k _ = [stringUtf8 (show (unK1 k))]
+
+instance GToSExpression Par1 where
+  gtoSExpression _ _ = []
+
+instance ToSExpression t => GToSExpression (Rec1 t) where
+  gtoSExpression (Rec1 t) = pure . toSExpression t
+
+instance (Foldable f, GToSExpression g) => GToSExpression (f :.: g) where
+  gtoSExpression (Comp1 fs) n
+    | null fs   = [nl n <> pad n <> "[]"]
+    | otherwise = nl n <> pad n <> "[" : foldMap gtoSExpression fs (n + 1) <> ["]"]
```

Comment:
We show collections wrapped in square brackets. Special handling of the `null` case gives us `[]` instead of `[  ]` for empty optional/optional-repeated fields, which is more pleasing.

We don’t yet add field names, but when we do this will be important to ensure that we label the collection and not every individual element thereof. It’s also just really nice looking and easier to parse visually since it doesn’t leave you with unexplained gaps in the serialized tree for missing optional elements.